Fix distributed autograd initialization.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29069 Fix distributed autograd initialization.**

Distributed autograd was initialized after RPC and this would cause a
race in some scenarios where one node might have initialized distributed
autograd, calls backward() but other nodes have not initialized distributed
autograd yet.

Moving this before `_init_rpc` fixes the problem since `_init_rpc` implicitly
has a sync between processes via the store.

Differential Revision: [D18280875](https://our.internmc.facebook.com/intern/diff/D18280875/)

Commit:
Fix distributed autograd initialization.

Distributed autograd was initialized after RPC and this would cause a
race in some scenarios where one node might have initialized distributed
autograd, calls backward() but other nodes have not initialized distributed
autograd yet.

Moving this before `_init_rpc` fixes the problem since `_init_rpc` implicitly
has a sync between processes via the store.

Differential Revision: [D18280875](https://our.internmc.facebook.com/intern/diff/D18280875/)

[ghstack-poisoned]

pietern commented on torch/distributed/rpc/__init__.py:
```diff
@@ -50,6 +50,14 @@ def init_model_parallel(
         store, _, _ = next(rendezvous_iterator)
 
+        # Initialize autograd before RPC since _init_rpc guarantees all
+        # processes sync via the store. If we initialize autograd after rpc,
```
`RPC` and `rpc`

pietern reviewed (approved):
LGTM

pietern reviewed (changes requested):
CI is failing.

The RPC initialization tests may crash because you now end up initializing distributed autograd twice.

mrshenli commented:
This would close #29156 and #29212

mrshenli commented:
If #29157 is merged before this one, could you enable those two tests?

rohan-varma commented:
As @mrshenli mentioned, I went ahead and disabled some of these tests in https://github.com/pytorch/pytorch/pull/29157 while we fix them.

zhaojuanmao commented:
it will close #28932 as well

mrshenli commented:
Shall we get this landed soon?

mrshenli commented:
This will also close #29410

Commit:
Update on "Fix distributed autograd initialization."


Distributed autograd was initialized after RPC and this would cause a
race in some scenarios where one node might have initialized distributed
autograd, calls backward() but other nodes have not initialized distributed
autograd yet.

Moving this before `_init_rpc` fixes the problem since `_init_rpc` implicitly
has a sync between processes via the store.

Differential Revision: [D18280875](https://our.internmc.facebook.com/intern/diff/D18280875/)

[ghstack-poisoned]

mrshenli commented on test/rpc_test.py:
```diff
@@ -265,10 +265,13 @@ def test_register_rpc_backend_and_start_rpc_backend(
         with self.assertRaisesRegex(RuntimeError, "is not unique"):
-            rpc.init_model_parallel(
-                self_name="duplicate_name",
+            store, _, _ = next(torch.distributed.rendezvous(
```
why calling the lower level APIs instead of `init_model_parallel`?

pritamdamania87 replied:
With the current changes `init_model_parallel` throws a KeyError because the worker_name_to_id[self_name] lookup fails for invalid strings. I call the lower level API here to reproduce the errors we're looking for.

mrshenli replied:
I see, we will need to update these tests after we move `worker_name_to_id` to `backend_config` dict.

mrshenli reviewed (approved):
This LGTM. Just had a question regarding init_rpc.

Commit:
Update on "Fix distributed autograd initialization."


Distributed autograd was initialized after RPC and this would cause a
race in some scenarios where one node might have initialized distributed
autograd, calls backward() but other nodes have not initialized distributed
autograd yet.

Moving this before `_init_rpc` fixes the problem since `_init_rpc` implicitly
has a sync between processes via the store.

Differential Revision: [D18280875](https://our.internmc.facebook.com/intern/diff/D18280875/)

[ghstack-poisoned]
